src/include/function/gds/gds_frontier.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include <atomic>
+#include <mutex>
+
+#include "common/types/internal_id_t.h"
+
+using namespace kuzu::common;
+
+namespace kuzu {
+namespace graph {
+class Graph;
+} // namespace graph
+
+namespace function {
+
+// TODO(Reviewer): Is this the right place to put this?
+using lock_t = std::unique_lock<std::mutex>;
+
+class RangeFrontierMorsel {
+    friend class PathLengthsFrontiers;
+
+public:
+    RangeFrontierMorsel() {}
+
+    bool hasNextVertex() const { return nextOffset < endOffsetExclusive; }
+
+    nodeID_t getNextVertex() { return {nextOffset++, tableID}; }
+
+protected:
+    void initMorsel(table_id_t _tableID, offset_t _beginOffset, offset_t _endOffsetExclusive) {
+        tableID = _tableID;
+        beginOffset = _beginOffset;
+        endOffsetExclusive = _endOffsetExclusive;
+        nextOffset = beginOffset;
+    }
+
+private:
+    table_id_t tableID = UINT64_MAX;
```

Comment:
use INVALID_TABLE_ID and INVALID_OFFSET for initialization

Reply:
Done.